### PR TITLE
Improve versioning display with `--version`

### LINKED ISSUE #814: F3D version and VTK version are incorrect when not release
**Describe the bug**
F3D version and VTK version are incorrect

**To Reproduce**
 1. Download the nightly
 2. Open the file using `f3d --version`

```
F3D 2.0.0

F3D - A fast and minimalist 3D viewer
Version: 2.0.0.
Build date: 2023-05-15 00:37:19.
Build system: Linux 64-bits.
Compiler: GNU 8.4.0.
External rendering module: ON.
Raytracing module: ON.
VTK version: 9.2.6 (build 20230224).
Copyright (C) 2019-2021 Kitware SAS.
Copyright (C) 2021-2023 Michael Migliore, Mathieu Westphal.
License BSD-3-Clause.
By Michael Migliore, Mathieu Westphal and Joachim Pouderoux.

```

**Expected behavior**

```
Version: 2.0.0-64-g....
VTK version: 9.2.6-...
```

### PR DESCRIPTION
**Pull request by mwestphal**
- Fix cmake logic to avoid including with full path
 - Use VTK_VERSION_FULL when possible to determine the VTK version to display
 - Rely on git-describe to create a F3D_VERSION_FULL
 - Fix #814 
 
 Before:
 ```
F3D 2.0.0

F3D - A fast and minimalist 3D viewer
Version: 2.0.0.
Build date: 2023-05-19 08:15:08.
Build system: Linux 64-bits.
Compiler: GNU 12.2.1.
External rendering module: OFF.
Raytracing module: OFF.
VTK version: 9.2.6 (build 20230427).
Copyright (C) 2019-2021 Kitware SAS.
Copyright (C) 2021-2023 Michael Migliore, Mathieu Westphal.
License BSD-3-Clause.
By Michael Migliore, Mathieu Westphal and Joachim Pouderoux.
 ```
 
 After:
 ```
 F3D 2.0.0

F3D - A fast and minimalist 3D viewer
Version: 2.0.0-69-g39a11ba6.
Build date: 2023-05-19 08:13:58.
Build system: Linux 64-bits.
Compiler: GNU 12.2.1.
External rendering module: OFF.
Raytracing module: OFF.
VTK version: 9.2.6-3279-ga3ba1fbbf3 (date: 20230427).
Copyright (C) 2019-2021 Kitware SAS.
Copyright (C) 2021-2023 Michael Migliore, Mathieu Westphal.
License BSD-3-Clause.
By Michael Migliore, Mathieu Westphal and Joachim Pouderoux.
 ```
 
 We could always show the full version for F3D, but I like it like that.

### CODE REVIEW

**Comment by mwestphal** [comment]
@snoyer please review, F3D will not lie about versions anymore :)

**Comment by snoyer** [comment]
> @snoyer please review, F3D will not lie about versions anymore :)

Generous of you to assume I know anything about cmake :) but I'm going to sit the review out.

Best I can do is test it...

nightly from sometime last month:
```
> /opt/f3d/bin/f3d --version
F3D 2.0.0

F3D - A fast and minimalist 3D viewer
Version: 2.0.0.
Build date: 2023-04-20 16:41:10.
Build system: Linux 64-bits.
Compiler: GNU 8.4.0.
External rendering module: ON.
Raytracing module: OFF.
VTK version: 9.2.6 (build 20230224).
Copyright (C) 2019-2021 Kitware SAS.
Copyright (C) 2021-2023 Michael Migliore, Mathieu Westphal.
License BSD-3-Clause.
By Michael Migliore, Mathieu Westphal and Joachim Pouderoux.
```

master:
```
make -j8 && bin/f3d --version
...
[100%] Built target f3d
F3D 2.0.0

F3D - A fast and minimalist 3D viewer
Version: 2.0.0.
Build date: 2023-05-19 09:10:57.
Build system: Linux 64-bits.
Compiler: GNU 13.0.1.
External rendering module: OFF.
Raytracing module: OFF.
VTK version: 9.2.5 (build 5).
Copyright (C) 2019-2021 Kitware SAS.
Copyright (C) 2021-2023 Michael Migliore, Mathieu Westphal.
License BSD-3-Clause.
By Michael Migliore, Mathieu Westphal and Joachim Pouderoux.
```

your branch:
```
> make -j8 && bin/f3d --version
CMake Warning at cmake/f3dVersion.cmake:48 (message):
  Version from git (1.3.1) disagrees with hard coded version (2.0.0).  Either
  update the git tags or main CMakeLists.txt.
Call Stack (most recent call first):
  CMakeLists.txt:39 (determine_version)
-- VTK 9.2.5 found
...
[100%] Built target f3d
F3D 1.3.1

F3D - A fast and minimalist 3D viewer
Version: 1.3.1-385-g39a11ba6.
Build date: 2023-05-19 09:01:15.
Build system: Linux 64-bits.
Compiler: GNU 13.1.1.
External rendering module: OFF.
Raytracing module: OFF.
VTK version: 9.2.5.
Copyright (C) 2019-2021 Kitware SAS.
Copyright (C) 2021-2023 Michael Migliore, Mathieu Westphal.
License BSD-3-Clause.
By Michael Migliore, Mathieu Westphal and Joachim Pouderoux.
```

Not super helpful observations:
- `git describe` reports weird stuff but maybe that's because it's a branch and not main? or do I need to explicitly update tags somehow? either way, F3D repeats what git says as intended, the wrong version is probably user error on my side.
- VTK build info got lost somehow?

**Comment by mwestphal** [comment]
- About the warning in `git describe`. Caused by an issue in the release process. The v2.0.0 tag had not been "annotated" I took care of that but you need to force update your tags locally with `git fetch --tags -f`. This issue will not appear on a fresh clone.

- VTK "build 5" is replaced by "date: 20230519" if it makes sense. In your case, since you are building against VTK 9.2.5 release, no need to add the date as it is a true release.

**Comment by snoyer** [comment]
All good then :+1:

File: CMakeLists.txt [hunk 1]
```diff
@@ -20,12 +20,23 @@ set(CMAKE_RUNTIME_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}/bin${output_postfix}")
 set(CMAKE_LIBRARY_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}/lib${output_postfix}")
 set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}/lib${output_postfix}")
 
+
```
Comment (Meakk) [inline CMakeLists.txt:23]: useless additional line

File: cmake/f3dVersion.cmake [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+# variables are left unchanged and status message is printed.
+#
+# This expect ${var_prefix}_MAJOR_VERSION, ${var_prefix}_MINOR_VERSION, ${var_prefix}_PATCH_VERSION
+# and ${var_prefix}_VERSION to be already set and will be used for checking when possible.
+#
+# Arguments are:
+#   source_dir : Source directory
+#   git_command : git executable
+#   var_prefix : prefix for variables e.g. "f3d".
+function(determine_version source_dir git_command var_prefix)
```
Comment (Meakk) [inline cmake/f3dVersion.cmake:25]: Like `f3d_embed_file`, should we prefix the CMake functions in this file by `f3d_`?

File: cmake/f3dVersion.cmake [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    endforeach()
+  else()
+    message(STATUS
+      "Could not use git to determine source version, using version ${${var_prefix}_VERSION}")
+  endif()
+endfunction()
+
+# Extracts components from a version string. See determine_version() for usage.
+function(extract_version_components version_string var_prefix)
+  string(REGEX MATCH "^v?(([0-9]+)\\.([0-9]+)\\.([0-9]+)-?(.*))$"
```
Comment (Meakk) [inline cmake/f3dVersion.cmake:63]: In our case, we are always prefixed by `v`

```suggestion
  string(REGEX MATCH "^v(([0-9]+)\\.([0-9]+)\\.([0-9]+)-?(.*))$"
```

**Review by Meakk — CHANGES_REQUESTED** [review]

**Review by Meakk — APPROVED** [review]

**Pull request merged**